### PR TITLE
Add license and doc

### PR DESCRIPTION
**Pull request by Sytten**
Mainly adds javadoc.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,11 +1,48 @@
...
+
+All external storage can be (and should be) wrapped in our asynchronous storage to avoid slowing down/stopping queries if external problems occurs with the database.
+
+## Getting Started
+#### Add SpillWay to your project pom
+
+```xml
+<dependency>
+    <groupId>com.coveo</groupId>
+    <artifactId>coveo-cloud-spillway</artifactId>
```
Comment (francistb) [inline README.md]: <artifactId>spillway</artifactId>

File: README.md [hunk 2]
```diff
@@ -1,11 +1,48 @@
+# SpillWay
 [![Build Status](https://travis-ci.org/coveo/spillway.svg?branch=master)](https://travis-ci.org/coveo/spillway)
 [![license](http://img.shields.io/badge/license-MIT-brightgreen.svg)](https://github.com/coveo/spillway/blob/master/LICENSE)
 [![Maven Central](https://maven-badges.herokuapp.com/maven-central/com.coveo/spillway/badge.svg)](https://maven-badges.herokuapp.com/maven-central/com.coveo/spillway)
 
-Sample usage
+## A distributed throttling solution
 
+SpillWay is an easy to use solution to add distributed throttling at the software level in your public API.
+This is particularly usefull if multiple services are running in different JVM.
```
Comment (GuiSim) [inline README.md]: useful, with only one `l`

File: src/main/java/com/coveo/spillway/storage/AsyncLimitUsageStorage.java [hunk 3]
```diff
@@ -16,6 +39,18 @@
 import java.util.concurrent.Executors;
 import java.util.stream.Collectors;
 
+/**
+ * An asynchronous implementation of {@link LimitUsageStorage}.
+ * </p>
+ * This storage internally uses a {@link InMemoryStorage} as cache and performs
+ * asynchronous calls to the distributed storage to share information.
+ * </p>
+ * This it particularly usefull when using a database over the network as
```
Comment (GuiSim) [inline src/main/java/com/coveo/spillway/storage/AsyncLimitUsageStorage.java]: useful

File: README.md [hunk 4]
```diff
@@ -1,11 +1,48 @@
...
+It is also possible to quickly to react when throlling happens with our built-in call-back mechanism.
+
+Storage currently supported:
+- In memory (for usage in the same JVM)
+- Redis
+
+All external storage can be (and should be) wrapped in our asynchronous storage to avoid slowing down/stopping queries if external problems occurs with the database.
+
+## Getting Started
+#### Add SpillWay to your project pom
```
Comment (GuiSim) [inline README.md]: Spillway, no capital W

File: README.md [hunk 5]
```diff
@@ -1,11 +1,48 @@
+# Spillway
 [![Build Status](https://travis-ci.org/coveo/spillway.svg?branch=master)](https://travis-ci.org/coveo/spillway)
 [![license](http://img.shields.io/badge/license-MIT-brightgreen.svg)](https://github.com/coveo/spillway/blob/master/LICENSE)
 [![Maven Central](https://maven-badges.herokuapp.com/maven-central/com.coveo/spillway/badge.svg)](https://maven-badges.herokuapp.com/maven-central/com.coveo/spillway)
 
-Sample usage
+## A distributed throttling solution
 
+Spillway is an easy to use solution to add distributed throttling at the software level in your public API.
+This is particularly useful if multiple services are running in different JVM.
```
Comment (malaporte) [inline README.md]: JVMs?

File: README.md [hunk 6]
```diff
@@ -1,11 +1,48 @@
...
+It is also possible to quickly to react when throlling happens with our built-in call-back mechanism.
+
+Storage currently supported:
```
Comment (malaporte) [inline README.md]: Storage backend?

File: README.md [hunk 7]
```diff
@@ -1,11 +1,48 @@
...
+- In memory (for usage in the same JVM)
```
Comment (malaporte) [inline README.md]: within the same

File: README.md [hunk 8]
```diff
@@ -1,11 +1,48 @@
...
+    <artifactId>spillway</artifactId>
+    <version>1.0.0</version>
+</dependency>
+```
+
+#### Usage
+###### Sample 1
 ```java
-    LimitUsageStorage storage = new RedisStorage("localhost");
+    LimitUsageStorage storage = new AsyncLimitUsageStorage(new RedisStorage("localhost"));
```
Comment (malaporte) [inline README.md]: I'm not sure I like the fact that the caller is responsible of creating an `AsyncLimitUsageStorage`. Will continue reading and maybe come back here.

File: README.md [hunk 9]
```diff
@@ -1,11 +1,48 @@
...
+    SpillwayFactory spillwayFactory = new SpillwayFactory(storage);
+
+    Limit<String> myLimit = LimitBuilder.of("myLimit").to(2).per(Duration.ofMinutes(1)).build();
+    Spillway<String> spillway = spillwayFactory.enforce("myResource", myLimit);
```
Comment (malaporte) [inline README.md]: hmm does this really needs to be generic?

File: pom.xml [hunk 10]
```diff
@@ -26,6 +26,11 @@
             <organization>Coveo</organization>
             <organizationUrl>http://github.com/coveo</organizationUrl>
         </developer>
+        <developer>
```
Comment (malaporte) [inline pom.xml]: Pour la postérité :p

**Comment by Sytten** [comment]
In understand why the test is failing and I will fix it.
Since the remove of old key is based on the time stamp, if the test is runned some time after the TIMESTAMP variable is declared, the limit will be removed automatically (and thus the test will fail).

File: pom.xml [hunk 11]
```diff
@@ -26,6 +26,11 @@
...
+            <name>Emile Fugulin</name>
+            <organization>Coveo</organization>
+            <organizationUrl>http://github.com/coveo</organizationUrl>
+        </developer>
     </developers>
 
     <scm>
```
Comment (Sytten) [inline pom.xml]: We should...

File: src/main/java/com/coveo/spillway/Spillway.java [hunk 12]
```diff
@@ -12,6 +43,18 @@
 import java.util.Map;
 import java.util.stream.Collectors;
 
+/**
+ * Contains methods to easily interact with the defined limits in the storage
+ * and test if the incomming query should be throttled.
```
Comment (malaporte) [inline src/main/java/com/coveo/spillway/Spillway.java]: incoming

File: pom.xml (hunk 10 above)
Comment (Sytten) [inline pom.xml]: Give me some credit fro writing all that javadoc :P

File: src/main/java/com/coveo/spillway/SpillwayFactory.java [hunk 13]
```diff
@@ -1,12 +1,54 @@
...
 package com.coveo.spillway;
 
+import com.coveo.spillway.limit.Limit;
+import com.coveo.spillway.limit.LimitBuilder;
+import com.coveo.spillway.storage.LimitUsageStorage;
+
+/**
+ * Factory to create {@link Spillway} objects using the specified storage method.
+ *
+ * @author Guillaume Simard
```
Comment (malaporte) [inline src/main/java/com/coveo/spillway/SpillwayFactory.java]: who's this guy? :p

File: README.md (hunk 9 above)
Comment (Sytten) [inline README.md]: Yes, see sample 2.

File: src/main/java/com/coveo/spillway/SpillwayFactory.java (hunk 13 above)
Comment (GuiSim) [inline src/main/java/com/coveo/spillway/SpillwayFactory.java]: 👻

File: src/main/java/com/coveo/spillway/limit/LimitBuilder.java [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+import java.util.function.Function;
+
+import com.coveo.spillway.trigger.LimitTrigger;
+import com.coveo.spillway.trigger.LimitTriggerCallback;
+import com.coveo.spillway.trigger.PercentageThresholdTrigger;
+import com.coveo.spillway.trigger.ValueThresholdTrigger;
+
+/**
+ * Contains helper methods to easily create {@link com.coveo.spillway.limit.Limit}
+ * Genereal usage is the following :
```
Comment (malaporte) [inline src/main/java/com/coveo/spillway/limit/LimitBuilder.java]: General

**Comment by Sytten** [comment]
we have a sonar report, PM if you want the link.

File: pom.xml (hunk 11 above)
Comment (Sytten) [inline pom.xml]: Done

File: README.md (hunk 9 above)
Comment (Sytten) [inline README.md]: By external interface you mean the Spillway class? Its basically a container of limits with some tools to interact with them.

**Pull request merged**